src/platform_impl/macos/mod.rs

```diff
@@ -3,100 +3,94 @@
+    fn create(
+        id: &TrayIconId,
+        attrs: &TrayIconAttributes,
+        mtm: MainThreadMarker,
+    ) -> crate::Result<(Retained<NSStatusItem>, Retained<TrayTarget>)> {
         let ns_status_item = unsafe {
-            let ns_status_item =
-                NSStatusBar::systemStatusBar(nil).statusItemWithLength_(NSVariableStatusItemLength);
-            let _: () = msg_send![ns_status_item, retain];
-            ns_status_item
+            NSStatusBar::systemStatusBar().statusItemWithLength(NSVariableStatusItemLength)
         };
 
         set_icon_for_ns_status_item_button(
-            ns_status_item,
+            &ns_status_item,
             attrs.icon.clone(),
             attrs.icon_is_template,
+            mtm,
         )?;
 
         if let Some(menu) = &attrs.menu {
             unsafe {
-                ns_status_item.setMenu_(menu.ns_menu() as _);
+                ns_status_item.setMenu((menu.ns_menu() as *const NSMenu).as_ref());
             }
         }
 
-        Self::set_tooltip_inner(ns_status_item, attrs.tooltip.clone())?;
-        Self::set_title_inner(ns_status_item, attrs.title.clone());
+        Self::set_tooltip_inner(&ns_status_item, attrs.tooltip.clone(), mtm)?;
+        Self::set_title_inner(&ns_status_item, attrs.title.clone(), mtm);
 
         let tray_target = unsafe {
-            let button = ns_status_item.button();
-
-            let frame: NSRect = msg_send![button, frame];
-
-            let target: id = msg_send![make_tray_target_class(), alloc];
-            let tray_target: id = msg_send![target, initWithFrame: frame];
-            let _: () = msg_send![tray_target, retain];
-            let _: () = msg_send![tray_target, setWantsLayer: YES];
-
-            let id = NSString::alloc(nil).init_str(&id.0);
-
-            (*tray_target).set_ivar(TRAY_ID, id);
-            (*tray_target).set_ivar(TRAY_STATUS_ITEM, ns_status_item);
-            (*tray_target).set_ivar(TRAY_MENU_ON_LEFT_CLICK, attrs.menu_on_left_click);
-            if let Some(menu) = &attrs.menu {
-                (*tray_target).set_ivar::<id>(TRAY_MENU, menu.ns_menu() as _);
-            }
-
-            let _: () = msg_send![button, addSubview: tray_target];
+            let button = ns_status_item.button(mtm).unwrap();
+
+            let frame = button.frame();
+
+            let target = mtm.alloc().set_ivars(TrayTargetIvars {
+                id: NSString::from_str(&id.0),
+                menu: RefCell::new(
+                    attrs
+                        .menu
+                        .as_deref()
+                        .and_then(|menu| Retained::retain(menu.ns_menu().cast::<NSMenu>())),
+                ),
+                status_item: ns_status_item.retain(),
+                menu_on_left_click: Cell::new(attrs.menu_on_left_click),
+            });
+            let tray_target: Retained<TrayTarget> =
+                msg_send_id![super(target), initWithFrame: frame];
+            tray_target.setWantsLayer(true);
```

Comment:
Just curious, what's the purpose of `setWantsLayer`?

Reply:
I'm not sure why it's used here, I just kept it as the previous implementation. Perhaps it's slightly more performant if one is not drawing into the view yourself?

Or are you're asking about the purpose of this in general? If so, then I'd recommend to read [the docs](https://developer.apple.com/documentation/appkit/nsview/1483695-wantslayer?language=objc), they're (for once) quite detailed.

Reply:
Oops, sorry. I thought it was a newly added line my mistake.